### ISSUE TITLE
NAP-6 Implementation Plan

### ISSUE THREAD

**Issue by DragaDoncila** [Napari/napari#7012]
## 🧰 Task
This issue is here to track current state of implementation w.r.t NAP-6. 

### Other Related Considerations
- [x] npe2 PR adding a mapping from command IDs to associated menu items for each plugin
    - just needs review I think?
    - useful because we need both the command object and the menu contrib object without iterating
- [x] PR adding empty placeholders to menus
    - If you look at #7011 you'll notice the empty submenus are not great as a UI experience. #6950 was the first attempt to address this issue, but it was ultimately implemented in #7038
    - [ ] The code needs testing and the keys should be moved off the layerlist context
- [ ] On contexts, keys, and context hierarchies

### Core NAP-6

- [x] Add `Layers -> Layer Type` and `Layers -> Data` submenus for conversions
    - not only can we populate with our own stuff, but @jni and I found stuff that could go in there from Tools menu
    - [x] update NAP
- [x] bring out stuff from `Layers -> Edit` into top `Layers`
    - Peter, Juan and I spoke about this at North America meeting and we felt that having top `Edit` and  `Layers -> Edit` is weird, but if we just move them out we can avoid this issue.
    - [x] NAP
- [x] top level Layers not contributable
    - [x] NAP
    - [x] maybe need to update the bit about submenu
- [x] name of plugin should be shown in menu next to action name
    - [x] pure commands need plugin name

This MVP version of NAP-6 will be released in v0.5.0.

- [ ] Add top level edit menu
    - [ ] update NAP
- [ ] Distribute layer context actions among Edit and Layers menus
    - [ ] update NAP
- [ ] Add top-level Develop menu (replaces Debug menu)
	- [ ] Activated by a preference
	- [ ] update NAP 
- [ ] Order & grouping of items
    - [ ] napari always on top
    - [ ] plugin declared group and order respected 
- [ ]  Decide exactly what types may be used as type hints for dependency injection in a plugin command function (these are the names declared in the `app.injection_store.namespace`), and ensure we have providers for these types. Also document this.
    - [ ] We are in rough agreement that "pure" commands in the `Layers` menu should work on the selected layer (if they take a layer at all). This should be documented and made explicit in the code.
- [ ] Decide on which context keys we want to make available to plugin developers for use in `when` and `enablement` expressions, where we expose them, and how developers can add more.
    - [ ] Implement parsing and respecting `when` expressions for menu items
    - [ ] Implement parsing and respecting `enablement` expressions for commands
- [ ] Allow plugins to declare arbitrary submenus under their own plugin’s `self` menu
    - [ ] NAP update from plugin name to `self`
- [ ] Add icon/something to signal when a menu item is a widget, so users will know when clicking on something will open a widget vs run something (we can and potentially should ensure adherence on the napari side)
- [ ] Adding name of plugin to menu item titles
    - [ ] @jni thinks we should be stripping `napari` from the plugin name (even the display name??) but I have concerns about mangling plugin names 
    - [ ] currently the name of the plugin also appears when the action is shown in the `Plugins` menu, even though all actions are grouped under a submenu of the same name. This is visually quite annoying, so we should look at whether we can strip the plugin name in this case. Note that a single action can't have different titles depending on the menu it's part of, so we need to either implement this, or change the name of the qt object directly.
- [ ] A key goal for our menu contributions is that users are never surprised by the outcome of clicking on an action. Since actions can come from all sorts of sources and **may** have destructive effects, we want to provide as much information as possible to users about each menu item
    - [ ] Add `description` field to manifest - make it required??
        - [ ] does it live on the `command` or on the `MenuContribution`? 
    - [ ] Read `description` into `status tip` field of `Action`
        - [ ] is that a super weird place to put it? Will users ever look there? 
- [ ] Rather than using the full menu path for contributable menus, plugin developers should be able to specify a unique key, and we complete the full menu path for them
    - [ ] Decide on the unique keys
    - [ ] Implement unravelling
    - [ ] Backwards compatibility? 
    - [ ] NAP
- [ ] Icons
    - [ ] Look at whether icons are an option for disambiguating menu item source (rather than full name) 
    - [ ] Figure out how to use superqt.fonticons and whether this should be exposed to plugin devs
    - [ ] Do they work in menu bars?
    - [ ] Add docs and potential examples to plugin template
- [ ] sort plugins menu alphabetically (lifted from #6410 as we should be doing this anyway really) 
- [ ] test coverage

### Next Gen Manifest
- [ ] Decide on schema for next gen manifest
    - [ ] Address concerns in #6227
- [ ] Formalize which `Action` fields are exposed to users via the manifest, and which contributions they are exposed in. Related issue #6224
- [ ] Update iterator APIs to match the new manifest structure and how we iterate in `napari` when registering actions
    - [ ] Related issue #6326  

### Doc Updates

- [ ] everything from implementation plan
- [ ] pure commands in Layers should act on selected layer
- [ ] more detail on each discussion/decision in bottom summary
- [ ] …s for widgets not …s for commands
- [x]  open issue add todo list point from nap
- [ ] context menu?
	- now that we've removed it from contributables, need to maybe update first example or say something about removal
- [ ] add usage documentation
- [ ] document exported context keys
- [ ] expose MenuContribution guide in npe2
- [ ] add menu contributions to napari-plugin-template

**Comment by lucyleeow** [comment]
So I made  #6174 a while ago but I am happy to close. Feel free to migrate any items from that issue.

**Comment by DragaDoncila** [comment]
Totally forgot about that issue... will need to transfer some stuff over

**Comment by lucyleeow** [comment]
I think manifest v2 would probably address https://github.com/napari/napari/issues/6227 (at least further manifest problems can be in a new issue)

**Comment by adamltyson** [comment]
We (@brainglobe) are very interested in plugin submenus. I have a couple of questions about the future implementation of this that weren't clear to me from NAP-6:
1. Will it be possible to have an arbitrary number of submenu levels (within reason!)?
2. Will it be possible for plugins from >1 package to go under the same menu? Currently BrainGlobe has 8 plugins that have between 1-3 widget contributions. We would ideally like to structure these something like this:

```
Plugins
└── brainglobe
    ├── brainglobe-registration*
    ├── brainglobe-segmentation*
    ├── brainglobe-stitch*
    ├── brainmapper*
    ├── brainreg*
    ├── brainrender*
    │   ├── brainrender
    │   └── manage-atlases
    └── cellfinder*
        ├── curate
        ├── detect
        └── train
```

Currently everything with a `*` is a separate plugin and PyPI package. We could achieve what we need by removing the contributions from each package and creating a "metaplugin" that defined the contributions for every plugin. We'd rather not do this because it would lead to some users needing to install a few GB of dependencies for a basic task. Ideally as each plugin was installed, it would just end up under the `brainglobe` submenu.

TL;DR - can multiple plugins share the same plugin submenu?

**Comment by DragaDoncila** [comment]
@adamltyson thanks for your comment - great to hear you're wanting to use the plugin submenus! Apologies for some of the details missing from the NAP-6 doc. They're not in there because we're not sure on the exact implementation yet heh 😅 

> Will it be possible to have an arbitrary number of submenu levels (within reason!)?

Yep, that's the goal. The rough idea is that in your manifest you would have something like this (for this example, assume all your commands are in one plugin):

```yaml
name: brainglobe-napari
display_name: Brainglobe

contributions:
  commands:
    # if this command is just under `Plugins -> Brainglobe` you could just not declare a menu item for it
    # but then you wouldn't be able to group or order it, it would just show up at the bottom of whatever commands you 
    # did declare menu items for
    - id: brainglobe-napari.register
      title: Register
      python_name: brainglobe_napari:register
    - id: brainglobe-napari.brainrender
      title: Brainrender from Submenu
      python_name: brainglobe_napari:brainrender

  menus:
    self:
      - submenu: self/brainrender
      - command: brainglobe-napari.register
         # would put this command at the top - not currently respected
         group: 0
    self/brainrender:
      - command: brainglobe-napari.brainrender
      - submenu: self/brainrender/deeper
    self/brainrender/deeper:
      - command: ...

  submenus:
    - id: self/brainrender
      label: brainrender
    - id: self/brainrender/deeper
      label: deeper submenu
```

Note that this is just the idea at the moment, and we'd appreciate all your feedback on whether this seems like what you'd want, or if the implementation is a little wonky. We're also looking at rearranging the manifest schema to reduce repetition, but that's a separate story.

> Will it be possible for plugins from >1 package to go under the same menu?

This is not currently in the implementation plan, but we have discussed before. 

For submenus **outside** the plugin's own menu (`self` menu), this is currently possible simply through shared submenu IDs. We haven't yet extended this to the `self` menu because we wouldn't want plugins to "hijack" other plugins' menus without consent 😆 . 

However, we were considering the idea of allowing a plugin to explicitly declare its own `self/*` submenus as "contributable", to enable exactly the kind of ecosystem you're referring to. In that world, you would edit your `brainglobe-napari` manifest to declare the `brainrender` menu as contributable e.g.

```yaml
name: brainglobe-napari
display_name: Brainglobe
...
submenus:
    - id: self/brainrender
      label: brainrender
      contributable: True
```

and then in your `napari-brainrender` plugin manifest you would have something like...

```yaml
name: napari-brainrender
display_name: Brainrender
menus:
  ...
  # use full plugin name of the plugin that declared this menu contributable
  brainglobe-napari/brainrender:
    - command: napari-brainrender.brainrender
```

How does that sound as a proposal? Feel free to tear it apart if you don't like it, the more we hear from you the more useful we can make this!

PS - brainrender is not an easy word to type 😆

**Comment by adamltyson** [comment]
Thanks for your reply @DragaDoncila. I think the hierarchy of submenus is exactly what we need.

Re "sharing" a submenu across plugins. This would be very valuable for us. We have a growing number of interoperable plugins, including some built by others (but built on top of BrainGlobe). It would be great be able to declare the `BrainGlobe` label as contributable, and ideally submenus too. 

My only question is whether the contributable menu could only be defined by one plugin? Ideally this could be defined by all plugins, and then any that were installed ended up in the same menu. 

> we wouldn't want plugins to "hijack" other plugins' menus without consent 

Personally I'm not so bothered about this. One of the side effects of open source, is that people may do things you don't love. It also wouldn't be too onerous to contact plugin devs if we thought that they were using a plugin menu in a manner confusing to users. If anybody did it maliciously, then the plugin could be blocked from the plugin manager.  Anyway, there's nothing to stop someone releasing a plugin called `napari-this-is-better-than-brainglobe-honest` anyway!

**Comment by DragaDoncila** [comment]
@adamltyson my apologies for the delay in getting back to you on this one!

> My only question is whether the contributable menu could only be defined by one plugin? Ideally this could be defined by all plugins, and then any that were installed ended up in the same menu.

Right now, this is how it works for external menus. So, using again the examples of `napari-brainrender` and `brainglobe-napari`, and assuming they both want to share a submenu labelled `Brainglobe Filters` under `Layers -> Filter`, you can achieve this with the following two manifests:

```yaml
name: brainglobe-napari
display_name: Brainglobe

contributions:
  ...
  menus:
    napari/layers/filter:
      - submenu: brainglobe_filters
    brainglobe_filters:
      - command: brainglobe_napari.some_filter
  # this plugin declares the ID and where it's placed
  submenus:
    - id: brainglobe_filters
      label: Brainglobe Filters
```

```yaml
name: napari-brainrender
display_name: Brainrender

contributions:
  ...
  menus:
    # this plugin just refers to the submenu by ID
    brainglobe_filters:
      - command: napari-brainrender.specific_filter
      ...
```

In the scenario above, one plugin declares the submenu and the other just uses it. But you could also declare the submenu in both manifests, and as long as the ID is the same and they're placed in the same `napari` menu, they'll behave as a single menu so the following manifests will achieve the same result as above.

```yaml
name: brainglobe-napari
display_name: Brainglobe

contributions:
  ...
  menus:
    napari/layers/filter:
      - submenu: brainglobe_filters
    brainglobe_filters:
      - command: brainglobe_napari.some_filter
  submenus:
    - id: brainglobe_filters
      label: Brainglobe Filters
```

```yaml
name: napari-brainrender
display_name: Brainrender

contributions:
  ...
  menus:
    # note the duplicate definition
    napari/layers/filter:
      - submenu: brainglobe_filters
    brainglobe_filters:
      - command: napari-brainrender.specific_filter
  # note the duplicate definition
  submenus:
    - id: brainglobe_filters
      label: Brainglobe Filters
```

If I'm understanding correctly, this second option is what you would like implemented? That is, both `brainglobe-napari` and `napari-brainrender` declare some submenu `brainglobe_filters` and this shows up under `Plugins -> Brainglobe Filters` rather than `Plugins -> Brainglobe -> Brainglobe Filters` and `Plugins -> Brainrender -> Brainglobe Filters` respectively? 

The `self` syntax I proposed above would result in the latter if **both** plugins declared a `self/brainglobe_filters` submenu. We hadn't yet considered/discussed opening the top-level `Plugins` menu itself for contribution, but I don't have any immediate objections to it! Curious what others think? @psobolewskiPhD? @jni?

If we were to open up the `Plugins` menu for contribution, then the example above with the `napari/layers/filter` menu would justwork™️  with `napari/plugins`. Then, we could possibly do away with the `self` syntax entirely, and plugins could just use `napari/plugins/plugin-name` when they wanted a submenu under their home menu at `Plugins -> Plugin Name`. They could still choose to declare any submenu under `Plugins -> Plugin Name` as contributable, but by default it wouldn't be.

How does all of that sound?

---

Couple of notes (both for myself and for whomever is following along) about what currently happens with submenu ID conflicts:

- if two plugins declare the same ID in the same location **with different labels**, the first plugin registered wins the label. Everything else works "as expected". We maybe should warn in this case?
- if two plugins declare the same ID **in different locations**, both menus get registered, and all commands registered under that ID (regardless of which plugin declared it) get placed in both menus. I'm not sure what we want to do about this 😬 Maybe just a warning is enough? Maybe it's not a big deal if we provide guidance about choosing IDs e.g. "start the ID with your plugin name if you don't want to risk a clash with communal IDs"?

Also cc'ing @isabela-pf on this one as it's relevant to our discussion from earlier today.

**Comment by adamltyson** [comment]
Hi @DragaDoncila, if I understand correctly, I think that's exactly what we need, thanks!

cc @alessandrofelder in case I've missed anything relevant to us.

**Comment by jni** [comment]
I'm +1 on allowing contributions to the Plugins menu.

For name clashes, I suppose I'm happy enough warning for now. When/if it becomes a problem, we can figure out something more principled.

**Comment by psobolewskiPhD** [comment]
I'm of two minds about the top-level Plugins menu editing. On one hand, I think letting plugins do things in `Plugins` is good: I think new users will tend to look for things there before exploring other menus? Particularly for widgets and things. So example 2 above would make life easier.
On the other hand, if there are both submenus for plugins (the current ones) and submenus *created by plugins* (the new ones, example 2 style above) then it gets to be confusing rather fast.

**Comment by DragaDoncila** [comment]
We chatted pretty in-depth about this at the Eurasia community meeting this week. The general feeling from everyone there was that we're happy to open up `napari/plugins` for contribution. The more we discussed it, the less concerned we were about the idea of plugins "hacking" other plugins' menus, but you definitely have a point @psobolewskiPhD about things potentially becoming messy.

Ultimately I think plugin collections are quite common, and I do think it makes sense for these plugins to be able to function as a unit. You could always get around it by having one "master" plugin that declares all your submenus, and then making all your other plugins depend on it, but that seems like a fair bit of boilerplate to support a pretty legitimate ask. 

At community meeting we also discussed having a big comment about things being experimental right now and still under review, so for now I'm definitely leaning towards opening up `napari/plugins` and walking that back if we need to.

We also discussed and came to consensus on the following items during community meeting:

- We settled on a plugin's home menu at `napari/plugins/my-plugin` being **contributable** by default, and allowing developers to toggle it off in the manifest. The key driver of this decision was that if a plugin is no longer maintained and someone wants to make a child plugin that depends on it, this new developer can now make full use of the plugin's home menu without needing a release from the original plugin. 
- We decided that ordering in the `napari/plugins` menu should be separate groups for built in stuff, then submenus declared by plugins alphabetically ordered, then plugin home menus alphabetically ordered. I'll need to update the NAP to mention this
- Folks weren't too concerned about the conflicts, because if submenus are meant for sharing, it's reasonable to assume plugin developers will come to consensus on things, and if the submenu isn't meant for sharing, you could always preface it with your own plugin name or any arbitrary string to ensure it's unique. When it comes to the conflict of different labels, we agreed that it's ok if it happens, but the name shown in the GUI should **be deterministic**, not liable to change between sessions. One suggestion was to use the modified date of the manifest file to ensure the oldest plugin always gets the label.

cc. @Czaki @brisvag @lucyleeow and @melonora who were there at community meeting, in case I've forgotten or misrepresented anything!

Curious what others think as well about these latest points!

**Comment by isabela-pf** [comment]
While contributable menu (or[ NAP-6](https://napari.org/stable/naps/6-contributable-menus.html)) work has long been underway, I’ve been asked to investigate some questions about how menus behave in wider software contexts and recommend how napari might organize and adjust the experience of menus to support contributable menus. I’ve also been told this is a time that other menu changes could/are being made (like adding the Layers menu) to improve menus while they are already being adjusted.

I reviewed menu items, groupings, and behavior from Imaris, Photoshop, Slack, Google Chrome, and macOS Finder for both image-specific use cases and more general industry standards. I compared these to napari viewer as it is now.

I’ve marked questions I would like replied to with a ❓so they stand out more clearly.

## Patterns

In reviewing other software, I found the following patterns of behavior.

- There is not repetition of menu items within menus.
- Repetition of commands found in menus does happen elsewhere in interfaces, like in context menus or buttons.
- `>` Indicates sub menus.
- Sub menus only go one level deep.
- `…` indicates opening additional interface before an action will be taken.
- The Edit menu seems relatively codified across applications. Photoshop seems to be the main exception, but only in that it adds a lot of menu items, it does not remove standard Edit commands like copy and pasting.
- Imaris uses icons in some menu items. Finder uses them even more sparingly, but does use them. This does not seem to be a standard use of menus, but there is some precedent.
- Menu items that are not default—whether user-contributed or sourced from another related application—are grouped in their own sections. Those sections appear in relevant menus, but they are always grouped and often at the bottom of that menu. I don’t think I found any instances of them being intermingled with default menu commands.
- Indications for whether a menu item will immediately initiate an action upon click or will put the user in a different mode is ultimately unclear. This is because menu items, unless they have a > for sub menus or … for additional interface, do seem to initiate action upon click by default. Buttons elsewhere in the interface that have a toggle on/off state when pressed are more likely to put users in a different mode than menu items will.

## My Recommendations

Based on the above patterns, the goals of contributable menus, and napari viewer’s current state, I recommend the following for napari’s menus going forward.

- Menu items should not repeat across different menus. Menu items belong to a single menu, but may appear elsewhere in the interface.
- Have standard use of `>`, `…`, and any other menu behavior indicators we decide on. Document them so they are maintained in future napari development. (Good news! It looks like napari viewer already follows this behavior! I do recommend keeping it.)
- Have the following menus: **Application (python)**, **File**, **Edit**, **View**, **Layer**, **Plugins**, **Tools**, **Window**, **Help**.
- ❓Tentatively, I want to recommend adding an Edit menu to napari. What menu items must be in an Edit menu seem to be **Undo**, **Redo**, **Cut**, **Copy**, **Paste**, **Show Clipboard**, **Clear Clipboard**; I recommend following that. The issue? I am not sure we have all these controls available in napari. Potentially I think it could be helpful to have this menu to show, for example, when a user is on a layer where Undo/Redo is possible they would be able to tell by what is active in the Edit menu. But if the menu items in Edit will nearly always be disabled because they are so rarely available, I don’t think we should have an Edit menu.
- Contributed menu items need to be in their own sections. Whether under a single menu (like Plugins) or in others they should not be intermingled with core (non-contributed) menu items.
- Contributed menus should be grouped in the lowest section of the menu. This is standard across applications I reviewed. 
- Menu items should activate a command, open a dialog or similar secondary UI for further interaction, or be a high-level toggle on/off. Menu items are not at the level of activating different cursor states or similar non-command functions.

As for indicating default or contributed menus

- Sectioning within a menu seems industry standard. It appeared in more applications I reviewed as a way to divide menu items that are core versus user added (whether as plugins or settings or something else). This would be my foremost recommendation.
- There is precedent for having icons in menus. It is difficult to make sure they are unique enough at that scale to be visible, and they will make menus wider which might create issues for people using napari with smaller screens.
- Sub menus by plugin, as napari has now, also may be worth keeping. It helps make clear what menu item comes from where in the case of troubleshooting or keeping track of what is needed to complete a task.
- No matter what direction is chosen, I do absolutely think what are core napari menus and what are user contributed menus needs to be marked within the viewer.

Contributed menu sections

To be honest, this one seems like a problem we are going to have a hard time solving without having errors or confusing items in the menus. I understand the desire (and already implemented backend) for it, but I am not sure how effective and clear this will be for users.

Ultimately I have more questions for the community before a recommendation.

- ❓How do plugin developers know there is a section for them to contribute to?
- ❓What do we do about conflicting names? Add a suffix by default? Alert the developer? Something else?
- ❓Is there a reason to not provide contributable menu sections as a part of core napari instead so there is a documented reference? Then plugin developers also have a central place to request more rather than having to search other plugins.

## What’s next

If there’s agreement on the menus I’ve proposed be in napari, I’ll mockup what those menus look like lined up and what menu items I recommend in each menu. 

I’ll also be keeping an eye out to reply to other questions, of course.

**Comment by lucyleeow** [comment]
> Is there a reason to not provide contributable menu sections as a part of core napari instead so there is a documented reference? Then plugin developers also have a central place to request more rather than having to search other plugins.

Sorry @isabela-pf , I didn't follow this. What is "contributable menu sections"?

**Comment by psobolewskiPhD** [comment]
> ❓Tentatively, I want to recommend adding an Edit menu to napari. What menu items must be in an Edit menu seem to be Undo, Redo, Cut, Copy, Paste, Show Clipboard, Clear Clipboard; I recommend following that. The issue? I am not sure we have all these controls available in napari. Potentially I think it could be helpful to have this menu to show, for example, when a user is on a layer where Undo/Redo is possible they would be able to tell by what is active in the Edit menu. But if the menu items in Edit will nearly always be disabled because they are so rarely available, I don’t think we should have an Edit menu.

I've very for an Edit menu (https://github.com/napari/napari/issues/5080)
I totally agree that it would help discoverability. might also drive feature requests.

**Comment by isabela-pf** [comment]
> Sorry @isabela-pf , I didn't follow this. What is "contributable menu sections"?

@lucyleeow no worries; I often found myself second-guessing my own terminology when writing this comment. Lots of words with too many potential meanings. 😆 I meant similar to what napari currently has under the new Layers menu

<img width="128" alt="" src="https://github.com/user-attachments/assets/3612ddc3-f50c-4cea-99ca-75b533be028f">

where menu items contributed by plugins will populate pre-grouped submenus. But for menus other than Layers (for example, if a plugin for opening certain image types wants to go under File with the other Open menu items) I was referring to grouping contributed menu items at the bottom of a menu as I've seen other applications do. Below a section like 

<img width="360" alt="" src="https://github.com/user-attachments/assets/d64ff237-dec4-49ca-b1a3-b38c9b139b95">

the line between Examples Gallery and Release notes or napari homepage and napari info on the Help menu. That's what I'm referring to as a section.

To summarize, I'm recommending:

- Keeping the submenus as they are in the Layers menu
- For other menus, if plugin authors decide to contribute there, grouping all contributed menu items in their own section at the bottom of that menu to clarify which menu items are core napari menu items and which are contributed from plugins.

Please let me know if that still didn't answer your question!

**Comment by DragaDoncila** [comment]
@isabela-pf thanks for all your work here, this is such great analysis and information to have ahead of the next steps of implementing NAP-6. 

I think everything you say seems reasonable. I agree with @psobolewskiPhD that it would be good to add an `Edit` menu, populated with whatever commands we currently support that are typically found in that menu. Seems like at this stage it would be mostly the clipboard stuff - copy, paste, show clipboard, clear clipboard. The `Labels` menu supports undo for painting, so we may want to add that as a menu item that's disabled when the active layer isn't `Labels`. However this menu item would be disabled most of the time so maybe that would actually be poor UI...

> Menu items should not repeat across different menus. Menu items belong to a single menu, but may appear elsewhere in the interface.

This would imply that plugin contributions that have associated menu contributions are no longer listed under `Plugins -> My Plugin`. I think this makes sense, I think it's weird for stuff to be duplicated. But I'm curious what others think, as well.

> How do plugin developers know there is a section for them to contribute to?

Each submenu they can contribute to will be listed in the plugin documentation (currently it's only listed under NAP-6). In terms of "sections" I tend to agree with you that if a developer wants to contribute to the top-level menu e.g. the `Help` menu (which is **not** open for contribution, but just for the sake of argument), those contributions should all be gathered and just added to the bottom. There shouldn't be an option for developers to, for example, slot something in under `napari homepage`, above the separator.

> What do we do about conflicting names? Add a suffix by default? Alert the developer? Something else?

We spoke at length about this at one of the community meetings, and summarised our thinking in [this comment](https://github.com/napari/napari/issues/7012#issuecomment-2333214011) (final bullet point). I think this might be a situation we want to play by ear and see how things develop as more people use the contributable menus. There is definitely a desire to share submenus across plugins, and I would rather err on the side of providing that freedom than implementing conflict resolution that ends up more confusing for both developers and users.

> Is there a reason to not provide contributable menu sections as a part of core napari instead so there is a documented reference? Then plugin developers also have a central place to request more rather than having to search other plugins.

I'm confused by this question, sorry. I think generally speaking, a menu/submenu should be contributable, but **not** a section within a default menu. This means that the sections within menus are entirely within napari's purview, and it's up to us to manage them. All plugin contributed stuff just goes to the bottom. There will be some complexity there with developers wanting to potentially group and order things across a "suite" of plugins, but this might also be something we want to wait on until we see more adoption.

**Comment by psobolewskiPhD** [comment]
> we may want to add that as a menu item that's disabled when the active layer isn't Labels. However this menu item would be disabled most of the time so maybe that would actually be poor UI...

Well, it's a way for people to know where it works, even if currently it's not in manu places.... might guilt trip us into implementing wider undo?

> This would imply that plugin contributions that have associated menu contributions are no longer listed under Plugins -> My Plugin. I think this makes sense, I think it's weird for stuff to be duplicated. But I'm curious what others think, as well.

I'd vote for duplication. One hand, the Plugins menu can become a mess, on the other hand, if I just installed a plugin because of <omg social media post omg>, I can quickly find TheWidget knowing the name of the plugin.

**Comment by lucyleeow** [comment]
Thanks @isabela-pf for explaining!

> I'm confused by this question, sorry. 

Yeah I think I was similarly confused about the 'contributable menu sections as a part of core napari instead so there is a documented reference". Do you mean a placeholder in the sections to say e.g., 'plugin contributions go here'?

**Comment by jni** [comment]
> Well, it's a way for people to know where it works, even if currently it's not in manu places.... might guilt trip us into implementing wider undo?

👆 I agree with all of this 😂

**Comment by isabela-pf** [comment]
I’m following up with my proposal for adjusting napari’s menus. I’ll break it up by each menu to make changes easier to find, but, as always, please let me know if you have any questions.

## Application (napari) menu

Keep menu as-is! 

## File menu

![](https://github.com/user-attachments/assets/ae5a7f0e-b177-4ac0-a8e0-0e08919b24cb)

Added
- Nothing

Removed
- New Layer (to be moved to Layers menu)

It makes sense why New Layer has been under the File menu near other “new x” menu items historically, but with a Layers menu added to napari I would advocate for keeping layer-related menu items there.

## Edit menu

![](https://github.com/user-attachments/assets/59814f83-314a-44ac-af97-9c23b804535e)

Added the Edit menu and all its contents, including
- Undo
- Redo
- Cut
- Copy
- Paste
- Show Clipboard
- Clear Clipboard

Removed
- Nothing

As has been discussed in previous comments, Undo and Redo in particular will be disabled on certain layers. I think it is valuable to add an Edit menu to communicate to users when Undo and Redo are available to them and when they are not. The other menu items are standard in other applications. I think Show Clipboard and Clear Clipboard are lower priority, but may prove helpful for troubleshooting.

## View menu

![](https://github.com/user-attachments/assets/b5606745-5dec-4c63-a7c8-c79c96eca0bb)

Added
- Nothing

Removed
- Toggle Activity Dock (to be moved to Window menu)

I would consider this change lower priority, but a more accurate use of menus. The activity dock is a contained UI area that can be shown and hidden just like all the menu items in Window. It stands out in the View menu that is more about changing visual elements on and around the canvas.

## Layers menu

![](https://github.com/user-attachments/assets/ad24fb38-b8a5-45c0-ab72-f73c9b813802)

Added
- New [Layer] (Moved from File)
- Duplicate Selected Layer
- Delete Selected Layer
- Rename Selected Layer
- Show All Layers
- Hide All Layers
- Show Selected Layer
- Hide Selected Layer
- Merge to Stack
- Selected Layer Opacity
- Selected Layer Blending

Removed (to be moved to Plugins menu)
- Visualize
- Annotate
- Data
- Layer Type
- Transform
- Measure
- Filter
- Register
- Project
- Segment
- Track
- Classify

I went through multiple options for how to group menu items in the Layers menu, and ultimately I think it is the one that would benefit most from having a more clearly defined goal. In my proposal, Layers’ goal is to group menu items for managing and adjusting layers based on napari’s core capabilities. This means that I have moved all plugin-populated menus (the contributable menus) out of it. Adding commands that currently exist only as buttons (like Delete Selected Layer, Rename Selected Layer, and so on) to the menus follows industry conventions, makes the commands findable via the Help menu search, and benefits any users for whom buttons may not be as easy to operate (like users unsure how to interpret icons, users not using a mouse, and users relying on assistive technology).. It makes actions that can be done to layers clearer from the start. To my knowledge, all the menu items I added here are already in napari, just not in the menus; please let me know if I’m wrong about any of those.

I will add that I think the section for layer controls that all layer types have—Selected Layer Opacity and Selected Layer Blending—is lower priority to add. Layer controls has it’s own share of complexities may prove too difficult or confusing. Still, I think having layer controls common across layer types as menus could benefit users in the same ways I listed in the last paragraph. 

## Plugins menu

![1 7](https://github.com/user-attachments/assets/130d11be-2add-47f7-bb78-d41538b6ffe7)

Added (moved from Layers)
- Visualize
- Annotate
- Data
- Layer Type
- Transform
- Measure
- Filter
- Register
- Project
- Segment
- Track
- Classify

Removed
- Nothing

Related to the Layers menu changes I proposed in the previous section, I would recommend moving the plugin-populated menus (the contributable menus) to the Plugins menu. Existing plugin menu items (Install/Uninstall Plugins… and Plugin Errors…) stay at the top of the list because they are populated by default napari. 

I believe this clarifies that these submenus and their menu items are all related to plugins, that they may not all act on layers, and makes their default Empty state make more sense. It’s less of a user asking “Well why can’t I Transform my selected layer? Is there something wrong with my layer? Is there something wrong with my napari installation to make Transform ‘Empty’?” and more of a user asking “I wonder what plugins I can install that will help me transform my selected layer?” 

Additionally, I think it supports the opportunity that contributable menus as a NAP aims for: providing a clear and well-scoped space for plugin developers to add menu items to napari without having to tackle napari core changes and without having to make sure every user can tell the difference between the default napari options and the plugin developer’s efforts. This all works without any need for additional fancy design (like icons or additional labels) in the submenus. 

![](https://github.com/user-attachments/assets/6f0e110a-a1b2-47e7-a515-16fd3c56f481)

## Window menu

![](https://github.com/user-attachments/assets/20941c5e-d313-495c-8cf4-b41d1c31b982)

Added (moved from View)
- [Toggle] Activity Dock

Removed
- Nothing

I recommend the Window menu stay almost exactly the same. In an ideal world, I would prefer Toggle Activity Dock appear in Window instead of View because the activity dock has more in common with the show/hide (or toggle) menu items in Window than the items in View, but that is lower priority in terms of changes to the menus.

## Help menu

![](https://github.com/user-attachments/assets/ad881835-7284-49c7-97bd-bdc9c2baf38b)

Added
- Nothing

Removed
- napari info (in favor of About napari in the napari menu)

I do consider this change lower priority. Right now both About napari and napari info exist in the default napari menus. They open the same dialog with the same information. These are duplicate commands with different names. About napari in the napari menu is the standard location and naming, so I would choose to keep that one and remove its duplicates.

**Comment by psobolewskiPhD** [comment]
The big shakeup is moving the "function"/"what they do" menus from Layer to Plugins.
Right now Plugins literally containers the menu items by Plugin. 
@isabela-pf How much value is in being able to find a specific plugin menu item after installing in Plugins, without hunting through the "function" menus?

Regarding the smaller bigs:
- The macOS `napari info` in Help is vestigial -- it was always there and we refer to it in the docs, because windows/linux don't have a `napari` menu. We could drop it, but then we'd need to delineate in the docs where to look for that function by platform.
- I agree with moving the Activity dock to Window, this should be a very small lift -- I will make an issue for it.
- I also like the additions from the contextual menu to the Layers menu -- we just talked about this here https://github.com/napari/napari/issues/7429
- and I'm very pro `Edit` menu

**Comment by DragaDoncila** [comment]
Yeah I don't hate moving the "category" menus from Layers to Plugins. My main concerns with it though are:

1. There will be built-in items in there e.g. under `Transform`, `Data` and `Layer Type`
2. We still wanted to keep the option for plugins to customize their own `Plugins -> My Plugin` submenu e.g. `Plugins -> Brainglobe -> (whatever Brainglobe want to do)`. After moving the "category" menus to `Plugins`, I guess these would all be listed at the bottom? The issue of fragmenting a single plugin's contributions still remains though: if Brainglobe has something they want to put in `Transform`, **and** stuff they want to keep in `Plugins -> Brainglobe`, now there are multiple places where the user has to look to find all Brainglobe things. Still, I guess at least they're only looking within one menu?

Otherwise:

- I like having `Layers -> New Layer`. That's where I had it initially actually haha, but then I moved it to `File` with the other `New` stuff 😆 
- I'm +1 for moving `Activity Dock` and for `Edit` menu
